_drafts/learn-doobie-for-great-good.md

```diff
@@ -0,0 +1,961 @@
+[error] 
+[error]   scala> Get[Foo]
+[error] 
+[error] And find the missing instance and construct it as needed. Refer to Chapter 12
+[error] of the book of doobie for more information.
+[error]     sql"select name from actors".query[ActorName].to[List].transact(xa)
+```
+
+Fortunately, Doobie gives us all the tools to quickly create such type classes. The first method we can use is to **derive the `Get[ActorName]` and `Put[ActorName]` type classes from the same defined for the `String` type**:
+
+```scala
+object ActorName {
+  implicit val actorNameGet: Get[ActorName] = Get[String].map(ActorName(_))
+  implicit val actorNamePut: Put[ActorName] = Put[String].contramap(actorName => actorName.value)
+}
+```
+
+As we may imagine, the `map` method is defined in the `Functor[Get]` type class, whereas the `contramap` method is defined in the `Contravariant[Put]` type class:
+
+```scala
+// Doobie library's code
+trait GetInstances extends GetPlatform {
+  implicit val FunctorGet: Functor[Get] =
+    new Functor[Get] {
+      def map[A, B](fa: Get[A])(f: A => B): Get[B] =
+        fa.map(f)
+    }
+}
+
+trait PutInstances extends PutPlatform {
+  implicit val ContravariantPut: Contravariant[Put] =
+    new Contravariant[Put] {
+      def contramap[A, B](fa: Put[A])(f: B => A): Put[B] =
+        fa.contramap(f)
+    }
+}
+```
+
+Moreover, in the case of newtypes, we can simplify the definition of the `Get` and `Put` type classes, using the `deriving` method available in the newtype library:
+
+```scala
+@newtype case class ActorName(value: String)
+object ActorName {
+  implicit val actorNameGet: Get[ActorName] = deriving
+  implicit val actorNamePut: Put[ActorName] = deriving
+}
+```
+
+Since it's widespread to derive both type classes for a type, Doobie also defines the `Meta[A]` type class, which allows us to create with a single statement both the `Get` and the `Put` type classes for a type:
+
+```scala
+implicit val actorNameMeta: Meta[ActorName] = Meta[String].imap(ActorName(_))(_.value)
+```
+
+The first parameter of the `imap` method is the function that defines the `Get` instance, whereas the second parameter is the function that defines the `Put` instance. The `imap` function comes from the `Invariant[Meta]` type class defined in the Cats library.
+
+Again, since the newtype library defines the `deriving` method, we can simplify the definition of the `Meta` type class:
+
+```scala
+implicit val actorNameMeta: Meta[ActorName] = deriving
+```
+
+**Doobie uses the `Get` and `Put` type classes only to manage single-column schema types**. In general, we need to map more than one column directly into a Scala class or into a tuple. For this reason, Doobie defines two more type classes, `Read[A]` and `Write[A]`, which can handle heterogeneous collections of columns.
+
+The `Read[A]` allows us to map a vector of schema types inside a Scala type. Vice versa, the `Write[A]` will enable us to map a Scala type into a vector of schema types.
```

Comment:
I know what you mean by "a vector of schema types", but for some reason the phrase doesn't seem clear enough to me. Maybe you can say that you can interact with more complex types instead of plain ints, strings etc.